allennlp/data/data_loaders/multi_process_data_loader.py

```diff
@@ -94,39 +94,37 @@ class MultiProcessDataLoader(DataLoader):
-        used to pass `Instance`s from worker processes to the consuming process when
-        `num_workers > 0`.
-
-        The default value should be reasonable for most use-cases.
+    !!! Note
+        In a typical AllenNLP configuration file, the `reader` and `data_path` parameters don't
+        get an entry under the "data_loader". The `reader` is constructed separately from
+        the corresponding `dataset_reader` params, and the `data_path` is taken from the
+        `train_data_path`, `validation_data_path`, or `test_data_path`.
 
-    instance_chunk_size: `int`, optional (default = `10`)
-        This determines how many `Instance`s are sent through the instance queue
-        at a time when `num_workers > 0`.
+    !!! Warning
+        Multiprocessing code in Python is complicated! Especially code that involves lower-level libraries
+        which may be spawning their own threads / processes. If you run into dead-locks while
+        using `num_workers > 0`, luckily there are two simple work-arounds which usually fix the issue.
 
-        The default value should be reasonable for most use-cases.
+        The first work-around is to disable parallelism for these low-level libraries.
+        For example, setting the environment variables `OMP_NUM_THREADS=1` and `TOKENIZERS_PARALLELISM=0`
+        will do so for PyTorch and Numpy (for CPU operations) and HugggingFace Tokenizers, respectively.
 
-    batch_queue_size: `int`, optional (default = `1000`)
-        This determines the size of the of the
-        [multiprocessing queue](https://docs.python.org/3.7/library/multiprocessing.html#multiprocessing.Queue)
-        used to pass `TensorDict` batches to the consuming process when `num_workers > 0`.
+        Alternatively, changing the `start_method` to "spawn" (when available, depending on your OS)
+        may fix your issues without disabling parallelism for other libraries.
 
-        The default value should be reasonable for most use-cases.
+        See [issue #4848](https://github.com/allenai/allennlp/issues/4848) for more info.
 
-    batch_chunk_size: `int`, optional (default = `10`)
-        This determines how many `TensorDict`s are sent through the batch queue at a time
-        when `num_workers > 0`.
+    !!! Warning
+        Another issue besides dead-locks that you could run into when using `num_workers > 0`
+        is running out of shared memory, since tensors are passed between processes
+        using shared memory, and some systems impose strict limits on the allowed size of shared
+        memory.
 
-        The default value should be reasonable for most use-cases.
+        Luckily there is also a simple work-around for this. Either decrease `max_instances_in_memory`
+        or increase your system's `ulimit`.
 
+        See [issue #4847](https://github.com/allenai/allennlp/issues/4847) for more info.
```

Comment:
Is there a way we can catch errors like this and make this message part of the error message?

Reply:
Not that I'm aware of. In some cases the process just hangs.

Reply:
😩